src/transform/src/analysis.rs

```diff
@@ -493,6 +493,97 @@ pub mod common {
+        pub fn new(global_mon_ids: BTreeSet<GlobalId>) -> Self {
+            Self { global_mon_ids }
+        }
+    }
+
+    impl Analysis for Monotonic {
+        type Value = bool;
+
+        fn derive(
+            &self,
+            expr: &MirRelationExpr,
+            index: usize,
+            results: &[Self::Value],
+            depends: &Derived,
+        ) -> Self::Value {
+            let mut child_results = depends
+                .children_of_rev(index, expr.children().count())
+                .map(|child| results[child])
+                .collect::<Vec<_>>();
+            child_results.reverse();
+            match expr {
+                MirRelationExpr::Get { id, .. } => match id {
+                    Id::Global(id) => self.global_mon_ids.contains(id),
+                    Id::Local(id) => *results
+                        .get(
+                            *depends
+                                .bindings()
+                                .get(id)
+                                .expect("Dependency info not found"),
+                        )
+                        .unwrap_or(&false),
+                },
+                MirRelationExpr::Project { .. } => child_results[0],
+                MirRelationExpr::Filter { predicates, .. } => {
+                    let is_monotonic = child_results[0];
+                    // Temporal predicates can introduce non-monotonicity, as they
+                    // can result in the future removal of records.
+                    // TODO: this could be improved to only restrict if upper bounds
+                    // are present, as temporal lower bounds only delay introduction.
+                    is_monotonic && !predicates.iter().any(|p| p.contains_temporal())
+                }
+                MirRelationExpr::Map { .. } => child_results[0],
+                MirRelationExpr::TopK { .. } => false,
+                MirRelationExpr::Reduce { aggregates, .. } => {
+                    // Reduce is monotonic iff its input is and it is a "distinct",
+                    // with no aggregate values; otherwise it may need to retract.
+                    child_results[0] && aggregates.is_empty()
+                }
+                MirRelationExpr::Union { .. } => {
+                    child_results.iter().fold(true, |acc, item| acc && *item)
+                }
+                MirRelationExpr::ArrangeBy { .. } => child_results[0],
+                MirRelationExpr::FlatMap { func, .. } => {
+                    child_results[0] && func.preserves_monotonicity()
+                }
+                MirRelationExpr::Join { .. } => {
+                    // If all inputs to the join are monotonic then so is the join.
+                    child_results.iter().fold(true, |acc, item| acc && *item)
+                }
+                MirRelationExpr::Threshold { .. } => child_results[0],
+                MirRelationExpr::Let { .. } | MirRelationExpr::LetRec { .. } => results[index - 1],
```

Comment:
We are inconsistent in our use of `child_results` v `results`. All of the `child_results[0]` could be `results[index-1]` I believe, or the other way around. Unless this is a load-bearing difference for a non-obvious (to me) reason, I propose we unify them up!